### PR TITLE
CBMC HTTP proof cleanup (depends on #1039)

### PR DESCRIPTION
**Pull request by markrtuttle**
This is a sequence of changes to the HTTP proofs intended to clean up the proofs in general, refactor the constructors in global_state_HTTP.c into allocators that do nothing but malloc memory and is_valid predicates that encode all of the assumptions we are making about the type (eg, that the malloc succeeded and the pointer is valid).  This refactoring has the advantage of recording all assumptions explicitly in the test harness itself, but the process of writing the is_valid predicates and testing that each of the remaining conjuncts was necessary led to weakening the assumptions (and thus strengthening the proofs).

We are also now testing for the full set of CBMC code checks (eg, testing for arithmetic overflow; testing for correctness even with arbitrary values written to uninitialized global state), and checking all configASSERTs embedded in the code.

We have also modified the test harness (how things like memcpy and snprintf are stubbed out) so that the proofs now build and run on both Linux and MacOS.

Warning: This pull request includes the two-line modification in PR 1039, so I recommend merging that one first and then this one.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json [hunk 1]
```diff
@@ -1,5 +1,10 @@
 {
   "ENTRY": "IotHttpsClient_Connect",
+
+  # This is the length of the user data after the header in a buffer.
+  # This can be set to any size.
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json]: If this is true, does it make sense to have e.g. three configurations of this proof: with `USER_DATA_SIZE` set to 0, 1, and 1000?

File: tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json [hunk 2]
```diff
@@ -1,5 +1,10 @@
...
+  "USER_DATA_SIZE": "1000",
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json]: There's no real need to set a variable up here if you're only using it in one place, I suggest erasing this line and writing `"USER_DATA_SIZE"="1000"` in the `"DEF"` array below.

File: tools/cbmc/proofs/HTTP/IotHttpsClient_Disconnect/Makefile.json [hunk 3]
```diff
@@ -1,5 +1,10 @@
 {
   "ENTRY": "IotHttpsClient_Disconnect",
+
+  # This is the length of the user data after the header in a buffer.
+  # This can be set to any size.
+  "USER_DATA_SIZE": "1000",
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_Disconnect/Makefile.json]: Same comments as above: suggest 3 different configurations, suggest not defining a variable here

File: tools/cbmc/proofs/HTTP/IotHttpsClient_ReadContentLength/Makefile.json [hunk 4]
```diff
@@ -1,5 +1,10 @@
 {
   "ENTRY": "IotHttpsClient_ReadContentLength",
+
+  # This is the length of the user data after the header in a buffer.
+  # This can be set to any size.
+  "USER_DATA_SIZE": "1000",
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_ReadContentLength/Makefile.json]: Same comments as above: suggest 3 different configurations, suggest not defining a variable here

File: tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/Makefile.json [hunk 5]
```diff
@@ -6,6 +6,10 @@
   "MAX_ACCEPTED_SIZE": 20,
   "STRNCPY_UNWIND": "__eval {MAX_ACCEPTED_SIZE} + 1",
 
+  # This is the length of the user data after the header in a buffer.
+  # This can be set to any size.
+  "USER_DATA_SIZE": "1000",
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/Makefile.json]: Same comments as above: suggest 3 different configurations, suggest not defining a variable here

File: tools/cbmc/proofs/HTTP/IotHttpsClient_ReadResponseStatus/Makefile.json [hunk 6]
```diff
@@ -1,5 +1,10 @@
 {
   "ENTRY": "IotHttpsClient_ReadResponseStatus",
+  
+  # This is the length of the user data after the header in a buffer.
+  # This can be set to any size.
+  "USER_DATA_SIZE": "1000",
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_ReadResponseStatus/Makefile.json]: Same comments as above: suggest 3 different configurations, suggest not defining a variable here

File: tools/cbmc/proofs/HTTP/IotHttpsClient_SendSync/Makefile.json [hunk 7]
```diff
@@ -1,9 +1,9 @@
 {
   "ENTRY": "HttpsClient_SendSync",
 
-  # This comes from libraries/c_sdk/standard/https/src/private/iot_https_internal.h
-  # But can't this be removed by including this header in the test harness?
-  "HTTPS_MAX_CONTENT_LENGTH_LINE_LENGTH": "26",
+  # This is the length of the user data after the header in a buffer.
+  # This can be set to any size.
+  "USER_DATA_SIZE": "1000",
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_SendSync/Makefile.json]: Same comments as above: suggest 3 different configurations, suggest not defining a variable here

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 8]
```diff
@@ -1,10 +1,10 @@
-#ifndef USER_DATA_SIZE
-#define USER_DATA_SIZE 1000
-#endif
+/****************************************************************/
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: Please could you not include these changes in future PRs, it's difficult for me to see whether you've actually changed the content of the comment

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 9]
```diff
@@ -38,47 +42,78 @@ typedef struct _connectionHandle
...
+  __CPROVER_assert(parser, "http_parser_execute parser nonnull");
+  __CPROVER_assert(settings, "http_parser_execute settings nonnull");
+  __CPROVER_assert(data, "http_parser_execute data nonnull");
+
   _httpsResponse_t *_httpsResponse = (_httpsResponse_t *)(parser->data);
   /* nondet_bool is required to get coverage. */
   _httpsResponse->foundHeaderField = nondet_bool();
   _httpsResponse->parserState = PARSER_STATE_BODY_COMPLETE;
+
+  size_t ret;
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: In the interest of non-CBMC people who read this: suggest
```c
size_t ret = nondet_size_t();
return ret;
```
(not sure if we have a `nondet_size_t()` implementation, but please add one if not)

File: tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json (hunk 1 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json]: Could we leave multiple configurations for after the proofs are completely done?  Unless I'm doing things wrong, they slow down the build-run-debug-commit cycle so much, and we still have so much work to do.

File: tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json (hunk 2 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json]: I prefer to leave this as it is.  This is a clean way to make it up front and center that there is an assumption being made.

File: tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json (hunk 1 above)
Comment (karkhaz) [reply tools/cbmc/proofs/HTTP/IotHttpsClient_Connect/Makefile.json]: Yes, agreed

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 8 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: I don't understand.  You appear to be seeing two adjacent changes made in two different commits in the PR that are being displayed together.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 9 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: Or just ```return nondet_size_t();``` . In fact, I could define nondet_MQTT_bool() and do the same thing in all of the stubs in global_state_HTTP.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 10]
```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
...
   }
   return pRequestHandle;
 }
+
+int is_valid_IotRequestHandle(IotHttpsRequestHandle_t pRequestHandle) {
+  int required =
+    __CPROVER_same_object(pRequestHandle->pHeaders,
+			  pRequestHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
+    __CPROVER_same_object(pRequestHandle->pHeaders,
```
Comment (sarenameas) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: The end needs to be the end of the buffer ((_reqHandle_t*)pRequestHandle)->data.
So it must be equal to (((_reqHandle_t*)pRequestHandle)->data + USER_DATA_SIZE)

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 11]
```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
...
+			  pRequestHandle->pHeadersEnd);
```
Comment (sarenameas) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: pRequestHandle->pHeadersEnd needs to be the end of the buffer ((_reqHandle_t*)pRequestHandle)->data.
So it must be equal to (((_reqHandle_t*)pRequestHandle)->data + USER_DATA_SIZE)

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 12]
```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
...
   return pResponseHandle;
 }
 
+int is_valid_IotResponseHandle(IotHttpsResponseHandle_t pResponseHandle) {
+  int required =
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersEnd);
```
Comment (sarenameas) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: Ok I looked at the bottom part and you are simply constraining pHeaders to be anywhere after pHeadersCur. So ignore my last comment.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 9 above)
Comment (sarenameas) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: That would make it easier to read. But looking at the rest of the code it is easy to see that these primitives by themselves become non-determinate. So I am ok with leaving as is.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 13]
```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
...
+IotHttpsConnectionHandle_t allocate_IotConnectionHandle () {
+  IotHttpsConnectionHandle_t pConnectionHandle =
+    safeMalloc(sizeof(_connHandle_t));
   if(pConnectionHandle) {
-    pConnectionHandle->pNetworkConnection = safeMalloc(sizeof(_connHandle_t));
-    pConnectionHandle->pNetworkInterface = newNetworkInterface();
+    // network connection just points to an allocated memory object
+    pConnectionHandle->pNetworkConnection = safeMalloc(1);
+    pConnectionHandle->pNetworkInterface = allocate_NetworkInterface();
+    // ???: Should reqQ initialization be an assumption?
```
Comment (sarenameas) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: It is a lot easier to make it empty. 
It sounds complicated to make it either empty or have items? That would require setting the links (nodes) correctly to wrap back around?

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 14]
```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
...
+      safeMalloc(pResponseHandle->readHeaderValueLength);
   }
   return pResponseHandle;
 }
 
+int is_valid_IotResponseHandle(IotHttpsResponseHandle_t pResponseHandle) {
+  int required =
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
```
Comment (sarenameas) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: Having pHeadersEnd be anywhere between pHeaders and pHeadersCur is correct if the application calls IotHttpsClient_InitializeRequest() before calling any other API.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c [hunk 15]
```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
...
-    pReqInfo->pHost = safeMalloc(hostNameLen);
+    pReqInfo->userBuffer.pBuffer = safeMalloc(pReqInfo->userBuffer.bufferLen);
+    pReqInfo->pHost = safeMalloc(pReqInfo->hostLen);
   }
   return pReqInfo;
 }
 
+int is_valid_IotRequestInfo(IotHttpsRequestInfo_t * pReqInfo) {
+  return
+    // ???: Should the global minimum size really be an upper bound?
```
Comment (sarenameas) [inline tools/cbmc/proofs/HTTP/global_state_HTTP.c]: No it should not actually. I did not catch that before.

Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: This is unusued, let's just delete the allocator and is_valid for this type.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 13 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: This constructor creates empty queues, but items are installed in the test harness itself.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 14 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: This was actually a question about cbmc itself, not about the ordering of pointers (we agree with your statement).

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 15 above)
Comment (sarenameas) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: That sounds good to me.

Comment (sarenameas) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: That sounds good to me.

File: tools/cbmc/proofs/HTTP/global_state_HTTP.c (hunk 14 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/global_state_HTTP.c]: The answer is "no", the code here is fine.

**Comment by markrtuttle** [comment]
Closing this PR and creating https://github.com/aws/amazon-freertos/pull/1047 that should be cleaner to merge into http.
